### PR TITLE
Implement UTXOSetValue for freezing

### PR DESCRIPTION
**Pull request by MichaelKim20**
Add freezing capability to UTXO #204 of part

UTXOSetValue is used by UTXOSet
To freeze UTXO, add the following information to UTXOSet:
The height of block.
The type of transaction.

UTXO 008

Block_Tx_UTXO

Transaction & UTXO in Freeze

### CODE REVIEW

**Comment by AndrejMitrovic** [comment]
This pull request introduces `findUTXOSetValue`, and then removes it in another commit. It should just be removed from the commit it was originally added in, and then the commit `Remove unused function UTXOSetValue.findUTXOSetValue` can be removed.

File: source/agora/consensus/data/UTXOSet.d [hunk 1]
```diff
@@ -27,6 +27,51 @@ import std.file;
 
 mixin AddLogger!();
 
+/// The structure of spendable transaction output
+public struct UTXOSetValue
+{
+    /// Height of the block to be activated
+    ulong height;
```
Comment (Geod24) [inline source/agora/consensus/data/UTXOSet.d]: This could use a more descriptive name.
I can suggest "height_lock", "frozen_until", "locked_until", "min_height"... But I'm open to suggestion @TrustHenry @AndrejMitrovic @MukeunKim

Comment (AndrejMitrovic) [reply source/agora/consensus/data/UTXOSet.d]: I suggest snake_case.

**Comment by Geod24** [comment]
> This pull request introduces findUTXOSetValue, and then removes it in another commit.b

Damn Github fix your commit ordering.

File: source/agora/consensus/data/UTXOSet.d (hunk 1 above)
Comment (TrustHenry) [reply source/agora/consensus/data/UTXOSet.d]: I suggest "height_lock".

File: source/agora/consensus/data/UTXOSet.d [hunk 2]
```diff
@@ -81,25 +126,116 @@ public class UTXOSet
...
+                    return;
+            }
+
+            foreach (const ref input; tx.inputs)
+            {
+                auto utxo_hash = getHash(input.previous, input.index);
+                this.utxo_db.remove(utxo_hash);
+            }
+
+            ulong actived_height = height + 1;
```
Comment (AndrejMitrovic) [inline source/agora/consensus/data/UTXOSet.d]: Why +1? The UTXO will be in the block at the provided height that's in the parameter, not height + 1.

File: source/agora/consensus/data/UTXOSet.d [hunk 3]
```diff
@@ -81,25 +126,116 @@ public class UTXOSet
...
-        foreach (const ref input; tx.inputs)
+        if (tx.type == TxType.Freeze)
         {
-            auto utxo_hash = getHash(input.previous, input.index);
-            this.utxo_db.remove(utxo_hash);
+            UTXOSetValue utxo_value;
+            foreach (const ref input; tx.inputs)
+            {
+                if (!this.getUTXOSetValue(input.previous, input.index, utxo_value))
+                    return;
```
Comment (AndrejMitrovic) [inline source/agora/consensus/data/UTXOSet.d]: These should `assert(0)`, not fail silently.

File: source/agora/consensus/data/UTXOSet.d [hunk 4]
```diff
@@ -81,25 +126,116 @@ public class UTXOSet
...
+                else if (utxo_value.type != TxType.Payment)
+                    return;
```
Comment (AndrejMitrovic) [inline source/agora/consensus/data/UTXOSet.d]: Same comment about assert.

File: source/agora/consensus/data/UTXOSet.d [hunk 5]
```diff
@@ -81,25 +126,116 @@ public class UTXOSet
...
+                    else if (input.previous == Hash.init)
+                        n_payment++;
+                }
+
+                if (n_payment == tx.inputs.length)
+                    tx_type = TxType.Payment;
+                else if (n_freeze == tx.inputs.length)
+                    tx_type = TxType.Freeze;
+                else
+                    return;
```
Comment (AndrejMitrovic) [inline source/agora/consensus/data/UTXOSet.d]: `assert(0)`

File: source/agora/consensus/data/UTXOSet.d (hunk 1 above)
Comment (MichaelKim20) [reply source/agora/consensus/data/UTXOSet.d]: I suggest "locked_until"~

File: source/agora/consensus/data/UTXOSet.d (hunk 2 above)
Comment (MichaelKim20) [reply source/agora/consensus/data/UTXOSet.d]: Next Block height

File: source/agora/consensus/data/UTXOSet.d (hunk 1 above)
Comment (AndrejMitrovic) [reply source/agora/consensus/data/UTXOSet.d]: I would go with @MukeunKim's suggestion, or maybe `unlock_height`

File: source/agora/consensus/data/UTXOSet.d (hunk 2 above)
Comment (MichaelKim20) [reply source/agora/consensus/data/UTXOSet.d]: I entered the available height, i.e. unlock height.
I wanted to specify what could be used in the next block, not in the current block.

File: source/agora/consensus/data/UTXOSet.d (hunk 1 above)
Comment (MichaelKim20) [reply source/agora/consensus/data/UTXOSet.d]: I like `unlock_height`

File: source/agora/consensus/data/UTXOSet.d (hunk 4 above)
Comment (MukeunKim) [reply source/agora/consensus/data/UTXOSet.d]: I use log.error()

File: source/agora/consensus/data/UTXOSet.d (hunk 3 above)
Comment (MukeunKim) [reply source/agora/consensus/data/UTXOSet.d]: I use log.error()

**Comment by MichaelKim20** [comment]
Ready fo Review

File: source/agora/consensus/data/UTXOSet.d (hunk 1 above)
Comment (TrustHenry) [reply source/agora/consensus/data/UTXOSet.d]: Me too `unlonk_height`  because  `height + 1;`

File: source/agora/consensus/data/UTXOSet.d (hunk 3 above)
Comment (MichaelKim20) [reply source/agora/consensus/data/UTXOSet.d]: Remove this part. I have confirmed that this is not necessary.

File: source/agora/node/Ledger.d [hunk 6]
```diff
@@ -542,7 +543,138 @@ unittest
...
+                tx_type,
+                last_txs,
+                GenesisTransaction);
+
+            txes.each!((tx)
+                {
+                    assert(ledger.acceptTransaction(tx) == is_valid);
+                });
+
+            if (is_valid)
```
Comment (AndrejMitrovic) [inline source/agora/node/Ledger.d]: `is_valid` is always true, because it is never set to false.

Comment (MichaelKim20) [reply source/agora/node/Ledger.d]: It's not used here, so we'll have to remove it

Comment (MichaelKim20) [reply source/agora/node/Ledger.d]: However, if an error occurs and the block is not created, then the values of these variables must not be changed.

Comment (MichaelKim20) [reply source/agora/node/Ledger.d]: Ready for review

**Comment by AndrejMitrovic** [comment]
Unless I'm missing something, this is what `updateUTXOCache` should look like:

```D
public void updateUTXOCache (const ref Transaction tx, ulong height) @safe
{
    ulong unlock_height = (tx.type == TxType.Freeze)
        ? height + 2016 : height + 1;

    foreach (const ref input; tx.inputs)
    {
        auto utxo_hash = getHash(input.previous, input.index);
        this.utxo_db.remove(utxo_hash);
    }

    Hash tx_hash = tx.hashFull();
    foreach (idx, output; tx.outputs)
    {
        auto utxo_hash = getHash(tx_hash, idx);
        const UTXOSetValue value = UTXOSetValue(unlock_height, tx.type, output);
        this.utxo_db[utxo_hash] = value;
    }
}
```

And also, this code passes the test-suite you wrote.

The validation rules for payment => freeze => melting should live in `isInvalidReason`, and not be scattered across multiple functions. `updateUTXOCache()`'s only purpose is to Remove spent UTXOs, and add new ones.

**Comment by AndrejMitrovic** [comment]
My previous comment was wrong with regards to unfreezing. Let me see if it can still be simplified..

File: source/agora/consensus/data/UTXOSet.d [hunk 7]
```diff
@@ -81,25 +126,105 @@ public class UTXOSet
...
+                        if (utxo_value.type == TxType.Payment)
+                            n_payment++;
+                        else
+                            n_freeze++;
+                    }
+                    else if (input.previous == Hash.init)
+                        n_payment++;
+                }
+
+                if (n_freeze == tx.inputs.length)
```
Comment (AndrejMitrovic) [inline source/agora/consensus/data/UTXOSet.d]: This is the part I find a bit confusing. The new outputs (UTXOs) will be be melting only if **all** of the inputs are frozen UTXOs. But if one of the was a Payment, then all of the outputs will not be melting, they will be instantly melted?

**Comment by MichaelKim20** [comment]
### There are three more commit after that.
It is PR #326.

**Includes:**
Implement validation of transactions associated with freezing
Add test for freezing and melting while creating blocks

**Comment by MichaelKim20** [comment]
Ready for review

File: source/agora/consensus/data/UTXOSet.d [hunk 8]
```diff
@@ -81,25 +126,89 @@ public class UTXOSet
...
+                    {
+                        tx_type = TxType.Freeze;
+                        break;
+                    }
+                }
+            }
+
+            foreach (const ref input; tx.inputs)
+            {
+                auto utxo_hash = getHash(input.previous, input.index);
```
Comment (AndrejMitrovic) [inline source/agora/consensus/data/UTXOSet.d]: This part seems to be the same for the freeze transactions too. Maybe it could be moved before the `if (tx.type == TxType.Freeze)` check so the code doesn't have to be duplicated.

**Comment by AndrejMitrovic** [comment]
I suggest rewriting these two functions as follows:

```D

/***************************************************************************

    Add all of a transaction's outputs to the Utxo set,
    and remove the spent outputs in the transaction from the set.

    Params:
        tx = the transaction
        height = Height of the block where UTXO will be unlocked

***************************************************************************/

public void updateUTXOCache (const ref Transaction tx, ulong height) @safe
{
    import std.algorithm : any;

    foreach (const ref input; tx.inputs)
    {
        auto utxo_hash = getHash(input.previous, input.index);
        this.utxo_db.remove(utxo_hash);
    }

    // defaults to next block
    ulong unlock_height = height + 1;

    // for payments of frozen transactions, it will melt after 2016 blocks
    if (tx.type == TxType.Payment
        && tx.inputs.any!(input => this.getUTXOSetValue(
            input.previous, input.index).type == TxType.Freeze))
    {
        unlock_height = height + 2016;
    }

    Hash tx_hash = tx.hashFull();
    foreach (idx, output; tx.outputs)
    {
        auto utxo_hash = getHash(tx_hash, idx);
        auto utxo_value = UTXOSetValue(unlock_height, tx.type, output);
        this.utxo_db[utxo_hash] = utxo_value;
    }
}

/***************************************************************************

    get an UTXOSetValue in the UTXO set.

    Params:
        hash = the hash of transation
        index = the index of the output
        value = will contain the UTXOSetValue if found

    Return:
        Return true if the UTXO was found

***************************************************************************/

private UTXOSetValue getUTXOSetValue (Hash hash, size_t index)
    nothrow @safe
{
    auto utxo_hash = getHash(hash, index);

    UTXOSetValue value;
    if (this.utxo_db.find(utxo_hash, value))
        assert(0);  // the UTXO must exist
    return value;
}
```

It gets rid of code duplication. And also, `getUTXOSetValue` must return a valid value because `updateUTXOCache` is called after validation. We know at this point a previous UTXO must exist.

**Comment by MichaelKim20** [comment]
To use 'any!' I think it's better to change it.

**Comment by MichaelKim20** [comment]
Ready for review

**Review by Geod24 — APPROVED** [review]

**Comment by AndrejMitrovic** [comment]
The bug in my suggestion was the over-eager removal of inputs. So this can be fixed to this:

```D
public void updateUTXOCache (const ref Transaction tx, ulong height) @safe
{
    import std.algorithm : any;

    // defaults to next block
    ulong unlock_height = height + 1;

    // for payments of frozen transactions, it will melt after 2016 blocks
    if (tx.type == TxType.Payment
        && tx.inputs.any!(input => this.getUTXOSetValue(
            input.previous, input.index).type == TxType.Freeze))
    {
        unlock_height = height + 2016;
    }

    Hash tx_hash = tx.hashFull();
    foreach (idx, output; tx.outputs)
    {
        auto utxo_hash = getHash(tx_hash, idx);
        auto utxo_value = UTXOSetValue(unlock_height, tx.type, output);
        this.utxo_db[utxo_hash] = utxo_value;
    }

    foreach (const ref input; tx.inputs)
    {
        auto utxo_hash = getHash(input.previous, input.index);
        this.utxo_db.remove(utxo_hash);
    }
}
```

**Comment by AndrejMitrovic** [comment]
However, this assertion will fail:

```D
private UTXOSetValue getUTXOSetValue (Hash hash, size_t index)
    nothrow @safe
{
    auto utxo_hash = getHash(hash, index);

    UTXOSetValue value;
    if (this.utxo_db.find(utxo_hash, value))
        assert(0);  // assertion triggers
    return value;
}
```

That means either something with the tests is wrong, or the validation functions aren't catching an edge-case.

**Comment by MichaelKim20** [comment]
@AndrejMitrovic I fixed it. Thank you.

**Comment by MichaelKim20** [comment]
The full test code using `UTXSet.updateUTXOCache` is in PR #326.
So we had a hard time proceeding with the review here.
Anyway, thank you all.
